File: backend/src/main/java/com/campuslostfound/model/Report.java
package com.campuslostfound.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.sql.Timestamp;
import java.util.List;

@Entity
@Table(name = "reports")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Report {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    @NotBlank(message = "Title is required")
    @Size(min = 1, max = 100, message = "Title must be between 1 and 100 characters")
    private String title;

    @Column(columnDefinition = "TEXT")
    @NotBlank(message = "Description is required")
    @Size(max = 1000, message = "Description must not exceed 1000 characters")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false, foreignKey = @ForeignKey(name = "fk_report_user"))
    @NotNull(message = "User is required")
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "location_id", nullable = false, foreignKey = @ForeignKey(name = "fk_report_location"))
    @NotNull(message = "Location is required")
    private Location location;

    @Column(name = "created_at", nullable = false, updatable = false, insertable = false, columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP")
    private Timestamp createdAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @NotNull(message = "Status is required")
    private ReportStatus status = ReportStatus.OPEN;

    @OneToMany(mappedBy = "report", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<Item> items;
}

File: frontend/src/models/report.ts
import type { User } from "./user";
import type { Location } from "./location";
import type { Item } from "./item";

export interface Report {
  id: number;
  userId: number;
  locationId: number;
  createdAt: string;
  status: ReportStatus;
  type: ReportType;
  itemIds?: number[];
  user?: User;
  location?: Location;
  items?: Item[];
}

export interface ReportCreateData {
  userId: number;
  locationId: number;
  type: ReportType;
}

export interface ReportUpdateData {
  userId?: number;
  locationId?: number;
  status?: ReportStatus;
}

export interface ReportFilters {
  userId?: number;
  locationId?: number;
  status?: ReportStatus;
}

// Enum for report status - matches backend but converted to TypeScript
export enum ReportStatus {
  OPEN = "OPEN",
  RESOLVED = "RESOLVED",
}

export enum ReportType {
  LOST = "LOST",
  FOUND = "FOUND",
}
